peloton/update_project.py

```diff
@@ -0,0 +1,1371 @@
+        #  (And immediately after encoding the previous value in the query
+        #  condition, if we're in an updates_only loop).
+        last_update_loop_time = datetime.now()
+        issues = IssuesQuery(query_string, peloton_logins)
+        discussions = DiscussionsQuery(query_string, peloton_logins)
+        logging.debug(
+            f"    Query returned {len(issues.data_frame)} issues and "
+            f"{len(discussions.data_frame)} discussions."
+        )
+
+        issues_discussions = pd.concat(
+            [issues.data_frame, discussions.data_frame]
+        )
+
+        logging.info("    Data fetch complete.")
+
+        if not updates_only:
+            # Removes all remaining project items that do not correspond with issues
+            #  or discussions returned in the query. E.g. an issue closed so long ago
+            #  that today's query no longer includes it.
+
+            # This is only run once, during the initial loop that gets everything
+            #  we want to be in the project. Subsequent loops being limited to
+            #  recent updates would mark almost every project item for removal.
+
+            no_linked_id = project.data_frame[project.cols.linked_id].isnull()
+            orphaned_linked_ids = ~project.data_frame[project.cols.linked_id].isin(
+                issues_discussions.index
+            )
+            items_to_remove = project.data_frame[no_linked_id | orphaned_linked_ids]
+            log_data_frame_items(items_to_remove, "    Removing from project")
+            RemoveItemsMutation(items_to_remove, project.cols, issues.cols)
+
+        # Link via a custom linked_id field to allow shared logic between issues
+        #  and discussions (as opposed to using GitHub Projects' native linking
+        #  that is only available for issues).
+        # Not passing rsuffix or lsuffix means we know our column names (`.cols`)
+        #  are unique with no overlaps between DataFrames - we can keep using them.
+        # We reset_index() for this since we want to have the id_in_project in our
+        #  resulting DataFrame.
+        joined = project.data_frame.reset_index().join(
+            issues_discussions,
+            on=project.cols.linked_id,
+            how="right",
+        ).set_index(project.cols.linked_id)
+
+        # Indicates that no associated project item was found during Pandas joining.
+        not_in_project = joined[project.cols.id_in_project].isnull()
+
+        have_new_comments = ~not_in_project & (
+            # Check two conditions to account for possibly deleted comments.
+            (
+                joined[issues.cols.total_comments] !=
+                joined[project.cols.num_comments]
+            )
+            |
+            (
+                # Forced to do date comparison (not datetime) because that is the
+                #  max precision of project.cols.final_comment_time .
+                joined[issues.cols.final_comment_time].str[:10] !=
```

Comment:
@trexfeathers The `:10` magic is used a few times in the script ... is this computable? or at least a constant with a comment to give it meaning to the reader 👍 

I'm guessing that it's related to the specific way datetime is formatted, so if this changes does this break? Seems kinda coupled ...

Reply:
Good idea for a convenience. Should be perpetually safe since the ISO format is an international standard.

e311cc1